text/0000-read-all.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: read_exact and read_full
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's `Write` trait has `write_all`, which is a convenience method that calls
+`write` repeatedly to write an entire buffer. This proposal adds two similar
+convenience methods to the `Read` trait: `read_full` and `read_exact`.
+`read_full` calls `read` repeatedly until the buffer has been filled, EOF has
+been reached, or an error other than `Interrupted` occurs. `read_exact` is
+similar to `read_full`, except that reaching EOF before filling the buffer is
+considered an error.
+
+# Motivation
+
+The `read` method may return fewer bytes than requested, and may fail with an
+`Interrupted` error if a signal is received during the call. This requires
+programs wishing to fill a buffer to call `read` repeatedly in a loop. This is
+a very common need, and it would be nice if this functionality were provided in
+the standard library. Many C and Rust programs have the same need, and solve it
+in the same way. For example, Git has [`read_in_full`][git], which behaves like
+the proposed `read_full`, and the Rust byteorder crate has
+[`read_full`][byteorder], which behaves like the proposed `read_exact`.
+[git]: https://github.com/git/git/blob/16da57c7c6c1fe92b32645202dd19657a89dd67d/wrapper.c#L246
+[byteorder]: https://github.com/BurntSushi/byteorder/blob/2358ace61332e59f596c9006e1344c97295fdf72/src/new.rs#L184
+
+# Detailed design
+
+The following methods will be added to the `Read` trait:
+
+``` rust
+fn read_full(&mut self, buf: &mut [u8]) -> Result<usize>;
```

Comment:
@ArtemGr 

An `uninterrupted()` doesn't fulfill the use case of `read_full`, or even make it easier. There is nothing that prevents one from using `read_full` where it's what one needs and `read` in other places where one needs more control, and it will work 100% correctly and as you'd expect.

`read_full` and `read_exact` are simply convenience functions that implement common patterns for using `read`. That's all. One can still do composition the same way by creating a wrapper type and implementing the `read` function of the `Read` trait, just like always. The final consumer can then call `read_full` on the composed result, and it will just work. `read_exact` and `read_full` have no impact on the ability to compose readers whatsoever, just like `write_all` doesn't make it any harder to compose writers.

> Suppose there is a generic code using `read` which was written without taking interruptible I/O into consideration. You can't use it with an interruptible I/O stream.

This seems like an argument for `read_full` rather than against. The code in question would be depending on unstated assumptions about the reader, and forgetting to wrap a reader with `uninterrupted()` before passing it would result in runtime failures with no help from the type system. Alternatively, the code could call `uninterrupted()` itself, but it would still have to loop to deal with non-EOF short reads, and in case the case that the composed reader already had `uninterrupted()` as one of the layers, the reader would be doing redundant work, checking for `Interrupted` multiple times per read. With this proposal, however, the code could just call `read_full`, which would just work no matter what reader was passed.